### ISSUE TITLE
Crash on screen focus

### ISSUE THREAD

**Issue by vezult** [zellij-org/zellij#3183]
# 2. Issues with the Zellij UI / behavior / crash

## Issue description

I run zellij on a remote machine, accessed via ssh/mosh, using gnome-terminal locally. I switched to my terminal where zellij was running, and it crashed immediately (or else it had already crashed, and that was when I saw the message). I *may* have focused the  window by clicking on it with my mouse. Apart from that, I had not interacted with zellij at all, and it was working fine when I left it.

```
Error occurred in server:
  × Thread 'screen' panicked.
  ├─▶ Originating Thread(s)
  │     1. stdin_handler_thread: AcceptInput
  │     2. screen_thread: WriteCharacter
  │   
  ├─▶ At /home/vezult/.cargo/registry/src/index.crates.io-6f17d22bba15001f/zellij-server-0.39.2/src/lib.rs:817:18
  ╰─▶ Program terminates: a fatal error occured
      
      Caused by:
          0: failed to write to active terminal for client 2
          1: failed to write to pane with id Terminal(1)
          2: failed to find pane with id Terminal(1)
  help: If you are seeing this message, it means that something went wrong.
        
        -> To get additional information, check the log at: /tmp/zellij-1000/zellij-log/zellij.log
        -> To see a backtrace next time, reproduce the error with: RUST_BACKTRACE=1 zellij [...]
        -> To help us fix this, please open an issue: https://github.com/zellij-org/zellij/issues
```

I do not have any fixed layout configuration, however, it is likely that this was a "revived" session at some point, and thus restored from some auto-saved layout configuration.

I keep perhaps 8-10 zellij tabs open, some of which are split, into multiple panes, in an ad-hoc fashion. It is *likely* that the active tab at the time of this crash was split into multiple panes. I cannot say that for certain.

I run my zellij session 24/7, so it would have been fairly long running, with multiple ssh/mosh client disconnects and reconnects.

## Minimal reproduction

Unfortunately, I do not know how to reproduce this.

## Other relevant information

```
❯ uname -a
Linux zapazoid 6.1.0-18-amd64 #1 SMP PREEMPT_DYNAMIC Debian 6.1.76-1 (2024-02-01) x86_64 GNU/Linux

❯ zellij --version
zellij 0.39.2

❯ rustc --version
rustc 1.76.0 (07dca489a 2024-02-04)
```

[zellij-crash-2024-03-06.log](https://github.com/zellij-org/zellij/files/14513093/zellij-crash-2024-03-06.log)

I notice that there are other, similar reports:
  * #1758
  * #1821
  * #2244
  * #2481
  * #2255
 
All of these reference `AcceptInput` and `WriteCharacter` threads, and seem to reference a similar region of `zellij-server/src/lib.rs`. I've *not* taken the time to track these references across the various versions concerned, to see if they are actually concerning the same operation.

Many of those seem to reference layout. I have no explicit fixed layout. I do not know how session restoration works though. It's possible that specifies fixed values for pane sizes, etc.

**Comment by imsnif** [comment]
My apologies for the experience! These are getting more and more rare and I'll definitely get this fixed ASAP.

**Comment by vezult** [comment]
So this happened again. Same error displayed on the terminal, same line of code referenced, same version of zellij. However this time on a machine installed via a binary download, rather than via cargo, FWIW.

In this instance, I switched zellj tabs, and hit "q". I don't *think* I hit Alt-q (to quit), but it's possible I did, which could mean that this is some sort of racy thing happening on shutdown. Who knows.

I'm including the full log this time in case other, non-fatal errors might provide some helpful history/context. So it's not just the entries recording this crash specifically.

This crash specifically starts at: `2024-03-13 14:50:17.845`

[zellij-error-2024-03-13.log](https://github.com/zellij-org/zellij/files/14592909/zellij-error-2024-03-13.log)

**Comment by imsnif** [comment]
Hey @vezult - I just issued a fix for this in #3209. These sorts of issues happen due to race conditions in our internal messaging (eg. in the above case it was a new tab opening and you typing in a command before the tab's layout was applied). These races are usually resolved internally in a very short time (measured in milliseconds or even microseconds), but in some cases (eg. heavy load / resource utilization) can take a bit longer.

I solved this for now but logging these sorts of errors rather than crashing - the effect of which in the future will be that worst case - you might lose one or two keypresses when creating new tabs or switching between them in heavy load scenarios.

In the future I hope to add some sort of caching/retry mechanism so that at most you'll experience a brief delay - but I think for now this should greatly improve the experience and hopefully your confidence in the app.

Thanks for reporting all these and for trusting in a fix!
